### PR TITLE
Fix retags in the wake of new reply/note layout

### PR DESCRIPTION
**Pull request by blackjackkent**

### CODE REVIEW

**Comment by nightpool** [comment]
As we're in the middle of a staged rollout, all of these changes should fallback to the old classes and whatnot, right?

File: Extensions/retags.js [hunk 1]
```diff
@@ -27,7 +27,7 @@ XKit.extensions.retags = {
 		var inner, $el;
 		ms.forEach(function(m){
 			$el = $(m.addedNodes);
-			inner = $el.find('.note:not(.ui_post_badge)');
```
Comment (nightpool) [inline Extensions/retags.js]: Won't this break the activity page?

File: Extensions/retags.js [hunk 2]
```diff
@@ -68,26 +68,27 @@ XKit.extensions.retags = {
...
 				}
 				cls = 'is_retags';
-				$c = $t.find('.stage');
-				url = $c.find('.part_glass').attr('href');
+				$c = $t.find('.stage > div:last-child').last();
+				url = $t.find('.part_glass').attr('href');
 			// dashboard
 			} else if ($t.hasClass('notification') && !XKit.browser().mobile) {
-				$c = $t.find('.notification_sentence');
+				$c = $t.find('.notification_sentence').last();
```
Comment (nightpool) [inline Extensions/retags.js]: why were these added?

**Comment by nightpool** [comment]
getting a lot of this:

image

**Comment by nightpool** [comment]
when clicking from the "conversation" to the "rollup" view

File: Extensions/retags.js (hunk 1 above)
Comment (blackjackkent) [inline Extensions/retags.js]: No -- activity page still works. (If I'm understanding correctly what's happening here, this is only for the mutation observer detecting the note popover.)

File: Extensions/retags.js (hunk 2 above)
Comment (blackjackkent) [inline Extensions/retags.js]: Because (see below line 185) we're now pasting the retags content after `$c` instead of appending within it. (Flexbox style weirdness was making the append not work very well with the new design.)

File: Extensions/retags.js (hunk 1 above)
Comment (nightpool) [inline Extensions/retags.js]: No, this Mutation Observer runs on every element added to the body, and detects every "note" class. note:not(ui_post_badge) is a way of filtering out the ".note" icon class that tumblr users on the activity page

Comment (nightpool) [inline Extensions/retags.js]: A reminder that retags covers: notes on the activity page, notes on some tumblr themes, notes directly on the dashboard, and the notes popup.

File: Extensions/retags.js [hunk 3]
```diff
@@ -20,16 +20,24 @@ XKit.extensions.retags = {
 		} catch(e) {}
 		this.add_toggle();
 		this.observer.observe($('body')[0],{childList:true,subtree:true});
-		this.tag(this.selectors);
+		XKit.extensions.retags.tag(XKit.extensions.retags.selectors);
```
Comment (nightpool) [inline Extensions/retags.js]: why was this changed and lines 21:22 weren't?

File: Extensions/retags.js [hunk 4]
```diff
@@ -20,16 +20,24 @@ XKit.extensions.retags = {
...
+		var inner, innerFallback, $el;
+		var notesProcessed = false;
 		ms.forEach(function(m){
 			$el = $(m.addedNodes);
-			inner = $el.find('.note:not(.ui_post_badge)');
-			if (inner.length > 0) {
-				XKit.extensions.retags.tag(inner.filter(XKit.extensions.retags.selectors));
+			var nonPopoverNotes = $el.find('.note:not(.ui_post_badge)');
+
+			if ($el.hasClass('note-list')) {
```
Comment (nightpool) [inline Extensions/retags.js]: How does this additional code in the mutation observer affect performance? I feel like this is a pretty big change to make, given the frequency that this MO is called

File: Extensions/retags.js [hunk 5]
```diff
@@ -181,7 +217,12 @@ XKit.extensions.retags = {
 				});
 			}
 			$t.addClass(cls);
-			$c.append($retags);
+			if (!isPopover) {
+				$c.append($retags);
+				return;
+			}
+			$retags = "<li class='retags-wrapper'>" + $retags[0].outerHTML + "</li>";
```
Comment (nightpool) [inline Extensions/retags.js]: does this need to be an li? is there really no way we could just use CSS to make this work?

adding an "isPopover" parameter to this call feels like a code smell to me.

Comment (blackjackkent) [inline Extensions/retags.js]: I agree, and tried for a good long time to make it work with pure CSS and didn't have any luck; if someone else wants to take a crack at it they're more than welcome and maybe we can pull this flag back out.

File: Extensions/retags.js (hunk 3 above)
Comment (blackjackkent) [inline Extensions/retags.js]: Ah. No reason, I was fiddling around with other changes here and removed this line and then re-added it by hand without thinking about it. I'll put it back  to how it was for consistency's sake.

File: Extensions/retags.js (hunk 4 above)
Comment (blackjackkent) [inline Extensions/retags.js]: @nightpool I haven't had a chance to look at this again since life got crazy again (watching calendar for June), but I have been using this extension locally on my machine since I submitted the PR and haven't noticed any performance degradation, FWIW.

Comment (nightpool) [inline Extensions/retags.js]: yeah, me too. I think it broke the normal note stuff though?

**Comment by nightpool** [comment]
Still getting this:
image

steps to reproduce: open the notes of a post with tags on it. Reply to that post. close the notes popup, open the notes popup.

File: Extensions/retags.js [hunk 6]
```diff
@@ -20,16 +20,25 @@ XKit.extensions.retags = {
 		} catch(e) {}
 		this.add_toggle();
 		this.observer.observe($('body')[0],{childList:true,subtree:true});
-		this.tag(this.selectors);
+		XKit.extensions.retags.tag(XKit.extensions.retags.selectors);
 	},
 
 	observer: new MutationObserver(function(ms){
-		var inner, $el;
+		var inner, innerFallback, $el;
```
Comment (nightpool) [inline Extensions/retags.js]: inner and innerFallback are unused here

File: Extensions/retags.js (hunk 3 above)
Comment (nightpool) [inline Extensions/retags.js]: still needs a fix

File: Extensions/retags.js [hunk 7]
```diff
@@ -61,24 +70,52 @@ XKit.extensions.retags = {
...
+				   $c = $t;
+				   url = $c.find('.action').data('post-url');
 			// Activity
 			} else if ($t.hasClass('ui_note')) {
 				if ($t.find('.part_response').length) {
 					$t.addClass('is_response');
 				}
 				cls = 'is_retags';
-				$c = $t.find('.stage');
+				$c = $t;
```
Comment (nightpool) [inline Extensions/retags.js]: Why was this changed?

File: Extensions/retags.js [hunk 8]
```diff
@@ -61,24 +70,52 @@ XKit.extensions.retags = {
 		}
 	},
 
-	tag: function(el){
+	tagPopover: function(el){
 		$(el).each(function(){
 			var $t = $(this), cls, $c, url;
```
Comment (nightpool) [inline Extensions/retags.js]: these variables are excruciatingly badly named.

File: Extensions/retags.js [hunk 9]
```diff
@@ -98,15 +135,15 @@ XKit.extensions.retags = {
...
 				} else {
-					XKit.extensions.retags.request($t,cls,$c,cache,'https://api.tumblr.com/v2/blog/'+host+'/posts/info?id='+id+'&api_key='+XKit.extensions.retags.api_key);
+					XKit.extensions.retags.request($t,cls,$c,cache,'https://api.tumblr.com/v2/blog/'+host+'/posts/info?id='+id+'&api_key='+XKit.extensions.retags.api_key,false);
 				}
 			}
 		});
 	},
 
-	request: function($t,cls,$c,cache,url) {
+	request: function($t,cls,$c,cache,url,isPopover) {
```
Comment (nightpool) [inline Extensions/retags.js]: use spaces in function argument declaration

File: Extensions/retags.js [hunk 10]
```diff
@@ -116,10 +153,10 @@ XKit.extensions.retags = {
 					XKit.extensions.retags.clear_old_posts();
 				}
 				XKit.storage.set('retags', cache, tags);
-				XKit.extensions.retags.append($t,cls,$c,tags);
+				XKit.extensions.retags.append($t,cls,$c,tags,isPopover);
```
Comment (nightpool) [inline Extensions/retags.js]: more spaces

File: Extensions/retags.js [hunk 11]
```diff
@@ -181,7 +218,12 @@ XKit.extensions.retags = {
 				});
 			}
 			$t.addClass(cls);
-			$c.append($retags);
+			if (!isPopover) {
```
Comment (nightpool) [inline Extensions/retags.js]: Instead of propagating a isPopover function we can just return `$retags` and move the bottom code up into the tag and tagPopover functions

File: Extensions/retags.js [hunk 12]
```diff
@@ -19,28 +19,31 @@ XKit.extensions.retags = {
...
+		this.observer.observe($('body')[0],{
+			childList:true,
+			subtree:true
+		});
+		this.tag(this.selectors);
 	},
 
 	observer: new MutationObserver(function(ms){
-		var inner, innerFallback, $el;
+		var $baseMutatedElement;
```
Comment (nightpool) [inline Extensions/retags.js]: plural

File: Extensions/retags.js [hunk 13]
```diff
@@ -1,6 +1,6 @@
 //* TITLE       Retags **//
 //* DEVELOPER   new-xkit **//
-//* VERSION     1.0.7 **//
```
Comment (nightpool) [inline Extensions/retags.js]: should be a patch bump

File: Extensions/retags.js [hunk 14]
```diff
@@ -19,111 +19,168 @@ XKit.extensions.retags = {
 			this.blog_name = $('body').data('new-root').split('/').pop();
 		} catch(e) {}
 		this.add_toggle();
-		this.observer.observe($('body')[0],{childList:true,subtree:true});
+		this.observer.observe($('body')[0],{
```
Comment (nightpool) [inline Extensions/retags.js]: spaces between arguments, maybe?

File: Extensions/retags.js [hunk 15]
```diff
@@ -19,111 +19,168 @@ XKit.extensions.retags = {
...
-				XKit.extensions.retags.tag(inner.filter(XKit.extensions.retags.selectors));
+			$baseMutatedElements = $(m.addedNodes);
+			var nonPopoverNotes = $baseMutatedElements.find('.note:not(.ui_post_badge)');
+			if ($baseMutatedElements.hasClass('note-list')) {
+				//remove existing retags items (if tabbing between popover views)
+				$baseMutatedElements.find('.retags-wrapper').remove();
+				var elements = $baseMutatedElements.find('.post-activity-note');
+				notesProcessed = true;
+				XKit.extensions.retags.tag_popover(elements.filter(XKit.extensions.retags.selectors));
+			} else if ($baseMutatedElements.hasClass('post-activity-note') && !notesProcessed) {
```
Comment (nightpool) [inline Extensions/retags.js]: feels like the else if and the notesProcessed variable are redundant w/ each other.

File: Extensions/retags.js [hunk 16]
```diff
@@ -19,111 +19,168 @@ XKit.extensions.retags = {
...
-				XKit.extensions.retags.css_toggle.appendTo('head');
-				if (XKit.browser().mobile) {
-					XKit.extensions.retags.mobile_toggle.appendTo('head');
+	tag_popover: function(elements){
+		$(elements).each(function(){
+			var $element = $(this), retagClass, $target, url;
+			if ($element.find('div.retags').length) {
+				return false;
+			}
+			if ($element.hasClass('is_reply') || $element.hasClass('is_answer') || $element.hasClass('type_6')) {
```
Comment (nightpool) [inline Extensions/retags.js]: "type_6"? 

I would write this like $element.is(".is_reply, .is_answer, .type_6")

File: Extensions/retags.js [hunk 17]
```diff
@@ -19,111 +19,168 @@ XKit.extensions.retags = {
...
+				return;
+			}
+			// popover
+			if ($element.hasClass('note-item')) {
+				retagClass = 'with_commentary';
+				$target = $element;
+				var peepr_data = $target.find('.note-text-link').data('peepr');
+				url = '';
+				if (peepr_data) {
+					url = 'http://' + peepr_data.tumblelog.trim() + '.tumblr.com/post/' + peepr_data.postId.trim();
```
Comment (nightpool) [inline Extensions/retags.js]: making the url and then deconstructing it is pretty weird flow

File: Extensions/retags.js [hunk 18]
```diff
@@ -19,111 +19,168 @@ XKit.extensions.retags = {
...
+				var host = url[2], id = url[4],	cache = 'post_' + id;
+				if (XKit.storage.get('retags', cache, null) !== null) {
+					var retags = XKit.extensions.retags.build_tag_collection($element, retagClass, XKit.storage.get('retags', cache));
+					if (!retags) {
+						return;
+					}
+					retags = "<li class='retags-wrapper'>" + retags[0].outerHTML + "</li>";
+					$target.after(retags);
+				} else {
+					var requestUrl = 'https://api.tumblr.com/v2/blog/' + host + '/posts/info?id=' + id + '&api_key=' + XKit.extensions.retags.api_key;
```
Comment (nightpool) [inline Extensions/retags.js]: why pass the request url? the function should know how to construct it. just past the tumblog + post id

File: Extensions/retags.js [hunk 19]
```diff
@@ -19,111 +19,168 @@ XKit.extensions.retags = {
...
+					XKit.extensions.retags.request($element, retagClass, cache, requestUrl)
+						.then(function(retags) {
```
Comment (nightpool) [inline Extensions/retags.js]: The cache should only be used at one level of abstraction—either entirely inside this function or entirely outside of it. right now we mix the two, leading to a leaky abstraction and weird side-effects. I would recommend encapsulating the caching logic inside this function.

File: Extensions/retags.js [hunk 20]
```diff
@@ -19,111 +19,168 @@ XKit.extensions.retags = {
...
-		}
 	},
 
-	tag: function(el){
-		$(el).each(function(){
-			var $t = $(this), cls, $c, url;
-			if ($t.find('div.retags').length) {
+	tag: function(elements){
+		$(elements).each(function(){
+			var $element = $(this), retagClass, $target, url;
```
Comment (nightpool) [inline Extensions/retags.js]: having actual names is beautiful, thank you.

File: Extensions/retags.js [hunk 21]
```diff
@@ -19,111 +19,168 @@ XKit.extensions.retags = {
...
-	request: function($t,cls,$c,cache,url) {
-		GM_xmlhttpRequest({
-			method: 'GET',
-			url: url,
-			onload: function(data){
-				var tags = JSON.parse(data.responseText).response.posts[0].tags;
-				if (XKit.extensions.retags.count_cached_posts() > XKit.extensions.retags.POST_CACHE_CLEAR_THRESHOLD || XKit.storage.quota('retags') < 100) {
-					XKit.extensions.retags.clear_old_posts();
+	request: function($element, retagClass, cache, url) {
+		return new Promise(function(resolve, reject) {
```
Comment (nightpool) [inline Extensions/retags.js]: is there a reason you can't just use $.get()? that returns a promise, so you don't have to mix the onload/promise APIs

File: Extensions/retags.js [hunk 22]
```diff
@@ -19,111 +19,168 @@ XKit.extensions.retags = {
...
+			GM_xmlhttpRequest({
+				method: 'GET',
+				url: url,
+				onload: function(data){
+					var tags = JSON.parse(data.responseText).response.posts[0].tags;
+					if (XKit.extensions.retags.count_cached_posts() > XKit.extensions.retags.POST_CACHE_CLEAR_THRESHOLD || XKit.storage.quota('retags') < 100) {
```
Comment (nightpool) [inline Extensions/retags.js]: sufficiently complex that it should probably be a function.

File: Extensions/retags.js [hunk 23]
```diff
@@ -19,111 +19,168 @@ XKit.extensions.retags = {
...
+						XKit.extensions.retags.clear_old_posts();
+					}
+					XKit.storage.set('retags', cache, tags);
+					var retags = XKit.extensions.retags.build_tag_collection($element, retagClass, tags);
+					resolve(retags);
+				},
+				onerror: function(data){
+					var retags = XKit.extensions.retags.build_tag_collection($element, retagClass, 'ERROR: ' + data.status);
```
Comment (nightpool) [inline Extensions/retags.js]: is there a promise.finally kinda thing we can use here?

Comment (nightpool) [inline Extensions/retags.js]: seems like if we use jquery promises, we get `Deferred.finally()`

File: Extensions/retags.js (hunk 15 above)
Comment (blackjackkent) [inline Extensions/retags.js]: @nightpool `notesProcessed` is initialized outside the foreach loop at line 32. It's watching to see if that variable has been set at any point in the loop.

File: Extensions/retags.js (hunk 13 above)
Comment (blackjackkent) [inline Extensions/retags.js]: @nightpool 1.0.7 to 1.0.8 is a patch bump, no?

File: Extensions/retags.js (hunk 15 above)
Comment (nightpool) [inline Extensions/retags.js]: oh duh my b. we should move the var $baseMutatedElements into the loop for clarity then

File: Extensions/retags.js [hunk 24]
```diff
@@ -19,111 +19,168 @@ XKit.extensions.retags = {
...
+			if ($element.find('div.retags').length) {
 				return false;
 			}
-			// popover
-			if ($t.hasClass('note')) {
-				cls = 'with_commentary';
-				$c = $t;
-				url = $c.find('.action').data('post-url');
+			//we don't need to put tags on a reply, but we also don't need to hide it
+			if ($element.hasClass('is_reply') || $element.hasClass('is_answer') || $element.hasClass('type_6')) {
```
Comment (nightpool) [inline Extensions/retags.js]: use `is()` as above

File: Extensions/retags.js [hunk 25]
```diff
@@ -19,111 +19,168 @@ XKit.extensions.retags = {
...
+				$target = $element.find('.notification-wrapper');
+				url = $target.find('a').not('.notification-username').attr('href');
 			}
 			if (url) {
 				url = url.split('/');
-				var host = url[2], id = url[4], cache = 'post_'+id;
+				var host = url[2], id = url[4], cache = 'post_' + id;
 				if (XKit.storage.get('retags', cache, null) !== null) {
-					XKit.extensions.retags.append($t, cls, $c, XKit.storage.get('retags', cache));
+					var retags = XKit.extensions.retags.build_tag_collection($element, retagClass, XKit.storage.get('retags', cache));
```
Comment (nightpool) [inline Extensions/retags.js]: fix cache as above

File: Extensions/retags.js (hunk 13 above)
Comment (nightpool) [inline Extensions/retags.js]: oh duh my b. I meant to say minor

File: Extensions/retags.js (hunk 16 above)
Comment (blackjackkent) [inline Extensions/retags.js]: type_6 was referenced in a previous version of the code before I even got here; I'm not sure what it refers to but there seemed no harm in leaving it as a fallback. I like this rephrase, thanks.

File: Extensions/retags.js [hunk 26]
```diff
@@ -19,111 +19,157 @@ XKit.extensions.retags = {
...
+		//if it's already cached, get from cache
+		if (XKit.storage.get('retags', cache_label, null) !== null) {
+			var retags = XKit.extensions.retags.build_tag_collection($element, retagClass, XKit.storage.get('retags', cache_label));
+			deferred.resolve(retags);
+			return deferred;
+		}
+		//otherwise make API call
+		var tags = [];
+		$.get(url)
+			.done(function(data){
```
Comment (nightpool) [inline Extensions/retags.js]: use the standard promise API of .then and .catch rather then the old jquery specific API

File: Extensions/retags.js [hunk 27]
```diff
@@ -19,111 +19,154 @@ XKit.extensions.retags = {
...
+				function(data){
+					tags = data.response.posts[0].tags;
+					if (XKit.extensions.retags.should_clear_posts()) {
+						XKit.extensions.retags.clear_old_posts();
+					}
+					XKit.storage.set('retags', cache_label, tags);
+					return tags;
+				},
+				function(data) {
+					tags = 'ERROR: ' + data.status;
```
Comment (nightpool) [inline Extensions/retags.js]: this means that we'll never propagate the error back up to the caller, right? because we're in the "error" part of the chain? We should return a deferred.resolve here too, so that we call the "then" callbacks up above

Comment (nightpool) [inline Extensions/retags.js]: or we could just handle the error explicitly at the caller site, with a separate "catch" block

File: Extensions/retags.js [hunk 28]
```diff
@@ -19,111 +19,154 @@ XKit.extensions.retags = {
...
+		return $.get(url)
+			.then(
+				function(data){
+					tags = data.response.posts[0].tags;
+					if (XKit.extensions.retags.should_clear_posts()) {
+						XKit.extensions.retags.clear_old_posts();
+					}
+					XKit.storage.set('retags', cache_label, tags);
+					return tags;
+				},
```
Comment (nightpool) [inline Extensions/retags.js]: should be `.catch` rather then the deprecated multi-function syntax

File: Extensions/retags.js [hunk 29]
```diff
@@ -116,16 +124,24 @@ XKit.extensions.retags = {
...
 				url = url.split('/');
 				var host = url[2], id = url[4];
-				XKit.extensions.retags.request($element, retagClass, host, id)
+				XKit.extensions.retags.request(host, id)
 					.then(function(tags) {
-						var retags = XKit.extensions.retags.build_tag_collection($element, retagClass, tags);
-						$target.append(retags);
+						XKit.extensions.retags.append_tag($element, retagClass, $target, tags);
+					})
+					.fail(function(errorResponse) {
```
Comment (nightpool) [inline Extensions/retags.js]: catch

**Comment by nightpool** [comment]
tested, works for me

@homu d+

**Comment by nightpool** [comment]
@homu delegate+

I'd recommend being around for the deploy, in case things go wrong, but TBH with our "every 12 hours" roll out system its kinda hard to tell when people are going to get new code so you might just be better served to put it out now. your call

**Comment by blackjackkent** [comment]
@homu r+
